### PR TITLE
Narrower Linter to Project NOT Global

### PR DESCRIPTION
**Pull request by kensipe**
Signed-off-by: Ken Sipe <kensipe@gmail.com>

**Description of the change:**

**Motivation for the change:**

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [ ] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

**Comment by kensipe** [comment]
cc: @joelanford @estroz 
This is a switch from GOBIN reference of linter to a project specific version.  It moves but maintains the same linters as previously along with the same version of golangci-lint.  It still resulting in several cases of lint issues needing resolution.  I'm not sure it is resolved in an ideal way and could use guidance.

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    - goimports
+    - errcheck 
+    - dupl 
+    - unparam 
+    - golint 
+    - staticcheck
+    - unused
+    - gosimple
+    # missing to be added
+    # - lll (250), errcheck, gocyclo, gosec, scopelint, unconvert, stylecheck
```
Comment (camilamacedo86) [inline .golangci.yml:22]: Hi @kensipe, 

Really thank you for your contribution.  Following some points.  

The goal of have `lll` is not allow the line to be bigger than the page (default value) which is bad since requires to use the scroll to check the code. However, was decided that it would be removed. I do not see the point to add it with 250 characters at all in this way. So, IMO we should not add the lll(250).

Also, see [here](https://github.com/operator-framework/operator-sdk/blob/master/hack/tests/check-lint.sh#L47-L66) the current checks that has been used by the CI. In this way, the `errcheck` should be enabled as well. 

Then, it is missing ` --deadline 5m` as well.

File: .golangci.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+run:
+linters-settings:
+  errcheck:
+    check-type-assertions: true
+  lll:
+    line-length: 250
```
Comment (camilamacedo86) [inline .golangci.yml:28]:
```suggestion
```

The goal of have `lll` is not allow the line to be bigger than the page (default value) which is bad since requires to use the scroll to check the code. However, was decided that it would be removed. I do not see the point to add it with 250 characters at all in this way. So, IMO we should not add the lll(250).

File: .golangci.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  dupl:
+    threshold: 400
```
Comment (camilamacedo86) [inline .golangci.yml:30]:
```suggestion
```
The dupl is working fine with its default value current.

File: .golangci.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
+linters:
```
Comment (camilamacedo86) [inline .golangci.yml:1]: IMO : It will be == [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/.golangci.yml) but with the current checks. See:  [here](https://github.com/operator-framework/operator-sdk/blob/master/hack/tests/check-lint.sh#L47-L66)

File: internal/scorecard/tests/olm.go [hunk 5]
```diff
@@ -335,7 +343,10 @@ func isCRFromCRDApi(cr unstructured.Unstructured, crds []*apiextv1.CustomResourc
...
 			}
 			if cr.Object["status"] != nil {
-				status := cr.Object["status"].(map[string]interface{})
+				status, ok := cr.Object["status"].(map[string]interface{})
+				if !ok {
+					failed = true
+				}
 				for key := range status {
 					if _, ok := version.Schema.OpenAPIV3Schema.Properties["status"].Properties[key]; !ok {
 						failed = true
```
Comment (camilamacedo86) [inline internal/scorecard/tests/olm.go:352]: If you apply the same checks as suggested no changes should be required. 
Then you can do another pr to enable any other check that you would like with the fixes.

File: .golangci.yml (hunk 2 above)
Comment (estroz) [reply .golangci.yml:28]: We intentionally removed `lll`: #3362

Comment (jmrodri) [reply .golangci.yml:28]: IMO it should be 95 (80 really)

File: .golangci.yml (hunk 1 above)
Comment (kensipe) [reply .golangci.yml:22]: @camilamacedo86 
`--deadline`  has been deprecated for sometime... `timeout` is it's replacement.   
`timeout` is on line 88 of the makefile of this PR (with the same value as deadline previously had)

File: .golangci.yml (hunk 2 above)
Comment (kensipe) [reply .golangci.yml:28]: while `lll` was left in the config, it wasn't enabled... I will remove from config.  I didn't realize longer lengths were controversial...   I'll leave this article regarding [Torvalds 2020 view of 80 char limits.](https://www.theregister.com/2020/06/01/linux_5_7/#:~:text=Linux%20kernel%20overlord%20Linus%20Torvalds,a%20thing%20of%20the%20past.)

File: .golangci.yml (hunk 1 above)
Comment (kensipe) [reply .golangci.yml:22]: `errcheck` was enabled on line 14 of this file.  It was mistakenly added to the line 22 comment... removed now.

File: internal/scorecard/tests/olm.go (hunk 5 above)
Comment (kensipe) [reply internal/scorecard/tests/olm.go:352]: @camilamacedo86 I'm not sure what you want to do here... The previous golangci-lint was version `v1.27.0`.  If I switch back to that, there are 3 "errcheck" failures in this file.  I do not understand why this doesn't come up in the CI or on others machines, but these are valid errcheck failures for which I've provided a fix.  Based on no differences captured in linting between `v1.27.0` and `v.1.30.0`, it seemed reasonable to update to the latest.

At this point:
* we can go with better code that doesn't ignore error checks
* I can remove the checks and add annotations to ignore these as err checks
* we can close this PR and continue to change developers global toolchain 

open to other suggests as well...

**Comment by kensipe** [comment]
@camilamacedo86 this PR should be only the previous checks... there were some configurations for checks that were not enabled, which could have lead to some confusion (these have been removed).
This PR represents:
1.  Having golangci-lint in the the project toolchain and not manipulating the developers global toolchain.
2. An upgrade of golangci-lint from 1.27.0 to 1.30.0 (because there was no linting output differences)
3. A change of golangci-lint, to be configured in a project `.golangci-lint` file (with the same configurations)
4. A change from deprecated `deadline` to preferred `timeout` (which provides the same function)

It does include errcheck changes for which I can't for why they were not captured previous.  I should add that I just checked out `master` and `make lint` and it fails... but there are multiple reasons (which this pr fixes)

Reasons for concern over current linting:
1. it affects the developers global toolchain
2. it does a naive check to see if golangci-lint exists on the local system... a simple `command -v $1 &>/dev/null` in `is_installed` function in `common.sh`.  This is likely the cause of many issues... based on this alone what is likely is a developer will have a `v1.15.0` on their machine and it will never be upgraded to the projects expected `v1.27.0`.   Many developers will get many different results.  The CI if it does NOT have golangci-lint installed will get what we expect... but if the image has a older version, it will forever run on that older version... eventually leading to developer confusion.
3. from the `check-lint.sh` line 10 is the install of golangci-lint... running that on my machine now, it silently fails.   try running: `curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh| sh -s -- -b $(go env GOPATH)/bin v1.27.0`.   In order to write these details, I did a little debugging... the issue is the `|` after `install.sh` needs a space.  The script to install golangci-lint in `check-lint.sh` currently never can install the linter.

Based on this... I have to assume that whatever linting results developers are getting is based on older versions of lint... thus no one ever discovered linting differences.  I also have to assume that your CI images have a version of linter installed... otherwise it would always fail (it appears).  

This PR resolves all those issues

File: .golangci.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  auto-fix: false
```
Comment (camilamacedo86) [inline .golangci.yml:2]: Please, auto-fix is very helpful. So, we should not add this. 
Currently, we have the make lint-fix which allow us to use this feature.

File: .golangci.yml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    - goimports
+    - errcheck 
+    - dupl 
+    - unparam 
+    - golint 
+    - staticcheck
+    - unused
+    - gosimple
+    # missing to be added
+    # - lll (250), gocyclo, gosec, scopelint, unconvert, stylecheck
```
Comment (camilamacedo86) [inline .golangci.yml:22]:
```suggestion
```

Comment (camilamacedo86) [reply .golangci.yml:22]: I do not think that we should add this comment. We do not know if we would like or not to ad them in the future.

File: internal/scorecard/tests/olm.go (hunk 5 above)
Comment (camilamacedo86) [reply internal/scorecard/tests/olm.go:352]: We should have a PR with the change that starts to raise the issue and the fixes for that only. Otherwise, it is hard we verify that. It might be a new check that and then, we should discuss if the new check should or not be added, If the new check is beneficial or not, as if the changes/fixes are accurate or not and etc..

File: .golangci.yml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+run:
+linters-settings:
+  errcheck:
+    check-type-assertions: true
```
Comment (camilamacedo86) [inline .golangci.yml:26]: Why this check is required?

Comment (kensipe) [reply .golangci.yml]: this is the "extra" check that I usually run with that was capturing the lack of error handling in `olm.go`.   Removing this, removes the need for change.

File: .golangci.yml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    - ineffassign
+    - goconst
+    - goimports
+    - errcheck 
+    - dupl 
+    - unparam 
+    - golint 
+    - staticcheck
+    - unused
+    - gosimple
```
Comment (camilamacedo86) [inline .golangci.yml:19]: nit. could you add an space at the end of the file?

File: Makefile [hunk 10]
```diff
@@ -79,14 +80,12 @@ tidy: ## Update dependencies
 clean: ## Clean up the build artifacts
 	$(Q)rm -rf build
 
-lint-dev:  ## Run golangci-lint with all checks enabled (development purpose only)
-	./hack/tests/check-lint.sh dev
-
-lint-fix: ## Run golangci-lint automatically fix (development purpose only)
-	./hack/tests/check-lint.sh fix
-
 lint: ## Run golangci-lint with all checks enabled in the ci
```
Comment (camilamacedo86) [inline Makefile:83]:
```suggestion
lint: ## Install and run golangci-lint checks
```

Comment (camilamacedo86) [reply Makefile:83]: I think we need to change the comment as well.

File: .gitignore [hunk 11]
```diff
@@ -99,6 +99,7 @@ test/helm-operator/helm-operator
 images/scorecard-proxy/scorecard-proxy
 images/scorecard-test/scorecard-test
 images/scorecard-test-kuttl/scorecard-test-kuttl
+/hack/bin
```
Comment (camilamacedo86) [inline .gitignore:102]: wdyt about we use the bin dir that already exist for we are able to keep all centralized instead?

Comment (kensipe) [reply .gitignore:102]: that was my original intention... but there isn't a `bin` in this project.   And I didn't see a target or script that builds to that folder... did I miss something?   Happy to transition to that location.

Comment (camilamacedo86) [reply .gitignore:102]: We should not have the bin since it is in the gitignore. It is not a good practice to commit them. 
IMO it is better we still using for the bin/ dir in the root of the project for we do not end up have many bins directories.

Comment (estroz) [reply .gitignore:102]: Built binaries are built to `build/`, currently nothing is built to `bin/`. I'm fine with a top-level `bin/` directory, and adding it to the `.gitignore` so binaries are not committed.

File: .golangci.yml [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
```
Comment (camilamacedo86) [inline .golangci.yml:20]: Nit ; 
Screen Shot 2020-08-10 at 16 27 26

Comment (camilamacedo86) [reply .golangci.yml:20]: You need to remove 1 space.

Comment (kensipe) [reply .golangci.yml:20]: it is unclear what the "nit" is... I took an earlier review comment that you wanted a newline... which is there now.   What specifically would you like resolved?

Comment (kensipe) [reply .golangci.yml]: oops... I'm getting used to a new editor and it provided a space / tab... nit makes sense.  apologies

File: .gitignore (hunk 11 above)
Comment (kensipe) [reply .gitignore:102]: I'm not sure where the confuse is coming from... but I would never commit this binary... not sure why that was assumed.   It was suggested in this thread that we build to `bin`.  my comment was that there currently isn't a `bin` for this project but there is an entry in the `.gitignore` file to ignore it.  which we would want to do in this case.  The question was should I start using a `bin/`... looks like we have 2 votes for that as the solution.   1 sec

Comment (kensipe) [reply .gitignore:102]: what I was originally trying to clarify is if I may have missed `bin` being used... because I didn't see it.   I questioned it, because there was a reference to `bin` in the ignore file... but I couldn't see a reason for it.

Comment (camilamacedo86) [reply .gitignore:102]: Hi @kensipe, 

the suggestion here is to install the `golangci-lint` in the `bin/` instead of `/hack/bin`. The reason is: If we start to add bin directories in the subdirectories we can end up in a scenario where we have more than one `bin` directory which reduces the maintenance ability.

Comment (kensipe) [reply .gitignore:102]: @camilamacedo86 that is what was in the last push from several hours ago

**Comment by kensipe** [comment]
@camilamacedo86 that was completed already... is there something missing?

**Comment by kensipe** [comment]
ah... I left a `hack/bin` in ignore file... other than that we should be good.

File: .gitignore (hunk 11 above)
Comment (camilamacedo86) [inline .gitignore:102]:
```suggestion
```

So, could you please revert the change here?

File: .golangci.yml [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
```
Comment (camilamacedo86) [inline .golangci.yml:20]: Also, see that it still with the simple. 
The files should have one 1 space at the end. 

@kensipe ^

Comment (kensipe) [reply .golangci.yml:20]: good catch!

**Comment by kensipe** [comment]
thanks so much @camilamacedo86 for you work, patience and review!

File: Makefile [hunk 14]
```diff
@@ -79,14 +80,12 @@ tidy: ## Update dependencies
...
-	./hack/tests/check-lint.sh fix
-
-lint: ## Run golangci-lint with all checks enabled in the ci
-	./hack/tests/check-lint.sh ci
+lint: ## Install and run golangci-lint checks
+ifneq (${GOLANGCI_LINT_VER}, "$(shell ./bin/golangci-lint --version 2>/dev/null | cut -b 27-32)")
+	@echo "golangci-lint missing or not version '${GOLANGCI_LINT_VER}', downloading..."
+	curl -sSfL "https://raw.githubusercontent.com/golangci/golangci-lint/v${GOLANGCI_LINT_VER}/install.sh" | sh -s -- -b ./bin "v${GOLANGCI_LINT_VER}"
+endif
+	./bin/golangci-lint --timeout 5m run
```
Comment (camilamacedo86) [inline Makefile:88]:
```suggestion
GOLANGCI_LINT_VER = "1.30.0"

lint: setup-lint ## Install and run golangci-lint checks
	$(LINT) --timeout 5m run

setup-lint: 
ifeq (, $(shell which golangci-lint --version 2 >/dev/null| cut -b 27-32))
	@{ \
	set -e ;\
	mkdir -p bin ;\
	echo "donwloading and installing golangci-lint@${GOLANGCI_LINT_VER} ..." ;\
	curl -sSfL "https://raw.githubusercontent.com/golangci/golangci-lint/v${GOLANGCI_LINT_VER}/install.sh" | sh -s -- -b ./bin "v${GOLANGCI_LINT_VER}" ;\
	}
LINT=$(realpath ./bin/golangci-lint)
else
LINT=$(shell which golangci-lint)
endif
```

Comment (camilamacedo86) [reply Makefile:88]: After check it better, I think it should be as the targets that we scaffolded with the tool.

File: Makefile [hunk 15]
```diff
@@ -26,6 +26,7 @@ GO_BUILD_ARGS = \
     -X '$(REPO)/internal/version.KubernetesVersion=$(K8S_VERSION)' \
   " \
 
+GOLANGCI_LINT_VER = "1.30.0"
```
Comment (camilamacedo86) [inline Makefile:29]: Please add it on top with the file with the other vars.

**Comment by estroz** [comment]
This can be merged after v1.0.0 is released.

/hold

File: Makefile (hunk 15 above)
Comment (kensipe) [reply Makefile:29]: @camilamacedo86 it is very confusing to know what you mean here... this in the middle of a series of variables which appear to be grouped for varies reasons.  It does go with the groupings for images, images and archs below.  you specify "top" and it is hard to claim that this is part of the group of build vars... but maybe... I don't see those in any particular order... I'll position this after k8s_ver but I'll need more specific guidance is that doesn't work for ya.

File: Makefile (hunk 14 above)
Comment (kensipe) [reply Makefile:87]: I don't understand... it only downloads if someone invokes `make lint` (in which case it is needed).  It only downloads if the version is the version defined, it doesn't overwrite the global GOBIN and it works.  What is current on master does not work.  IF there is some additional features or desire, perhaps we can create an issue with details and create a future PR? 
I really can't understand the comment as written or what is intended.

Comment (camilamacedo86) [reply Makefile:87]: Following a few motivations for the suggestion:
- Following the project standard. (See the Makefile of the projects generated by the tool)
- In your implementation, it will always download the lint in the bin/ dir. In the suggestion, it ONLY downloads the lint if it is NOT be installed already in the local env. See that it not necessary needs to be in the bin/ dir of the project.
- Allow works with the lint installed outside of the project. 
- Maintenance ability

Comment (camilamacedo86) [reply Makefile:87]: I am closing this one since we already spoke about. Please, let feel free to ping me when you are able to address the nit change.
